server/manager/src/main/java/org/apache/accumulo/manager/upgrade/UpgradeProgressTracker.java

```diff
@@ -0,0 +1,162 @@
+        throws KeeperException, InterruptedException {
+      Objects.requireNonNull(context, "ServerContext must be supplied");
+      Preconditions.checkArgument(newVersion <= AccumuloDataVersion.get(),
+          "New version (%s) cannot be larger than current data version (%s)", newVersion,
+          AccumuloDataVersion.get());
+      Preconditions.checkArgument(newVersion > metadataVersion,
+          "New Metadata version (%s) must be greater than current version (%s)", newVersion,
+          metadataVersion);
+      Preconditions.checkArgument(newVersion <= zooKeeperVersion,
+          "New Metadata version (%s) expected to be <= ZooKeeper version (%s)", newVersion,
+          zooKeeperVersion);
+      Preconditions.checkArgument(newVersion <= rootVersion,
+          "New Metadata version (%s) expected to be <= Root version (%s)", newVersion, rootVersion);
+      metadataVersion = newVersion;
+      put(context, this);
+    }
+  }
+
+  private static String getZPath(ServerContext context) {
+    return context.getZooKeeperRoot() + Constants.ZUPGRADE_STATUS;
+  }
+
+  private static synchronized void put(ServerContext context, ComponentVersions cv)
+      throws KeeperException, InterruptedException {
+    Objects.requireNonNull(context, "ServerContext must be supplied");
+    Objects.requireNonNull(cv, "ComponentVersions object  must be supplied");
+    final String zpath = getZPath(context);
+    final ZooReaderWriter zrw = context.getZooSession().asReaderWriter();
+    zrw.sync(zpath);
+    if (!zrw.exists(zpath)) {
+      zrw.mkdirs(zpath);
+    }
+    zrw.putPersistentData(zpath, GSON.get().toJson(cv).getBytes(UTF_8), NodeExistsPolicy.OVERWRITE);
+  }
+
+  public static synchronized ComponentVersions get(ServerContext context)
+      throws KeeperException, InterruptedException {
+    final String zpath = getZPath(context);
+    final int currentVersion = AccumuloDataVersion.getCurrentVersion(context);
+    final ZooReaderWriter zrw = context.getZooSession().asReaderWriter();
+    zrw.sync(zpath);
+    if (!zrw.exists(zpath)) {
+      ComponentVersions cv = new ComponentVersions();
+      cv.zooKeeperVersion = currentVersion;
+      cv.rootVersion = currentVersion;
+      cv.metadataVersion = currentVersion;
+      put(context, cv);
+      return cv;
+    } else {
+      byte[] jsonData = zrw.getData(zpath);
+      return GSON.get().fromJson(new String(jsonData, UTF_8), ComponentVersions.class);
+    }
+  }
+
+  public static synchronized void upgradeComplete(ServerContext context)
+      throws KeeperException, InterruptedException {
+    // This should be updated prior to deleting the tracking data in zookeeper.
+    Preconditions
+        .checkState(AccumuloDataVersion.getCurrentVersion(context) == AccumuloDataVersion.get());
+    final String zpath = getZPath(context);
+    final ZooReaderWriter zrw = context.getZooSession().asReaderWriter();
+    zrw.sync(zpath);
+    zrw.recursiveDelete(zpath, NodeMissingPolicy.SKIP);
```

Comment:
You don't need to sync because write operations already require quorum.

```suggestion
    zrw.recursiveDelete(zpath, NodeMissingPolicy.SKIP);
```

Reply:
Addressed in 7d19a91